### DISCUSSION TITLE
Adding logging/monitoring before reaching limit

### DISCUSSION THREAD

**Discussion by dmitry-naumovich** [MarcGiffing/bucket4j-spring-boot-starter#285] [Q&A]
Hello, our application uses bucket4j-spring-boot-starter to implement throttling for some of our API-s which are used by external partners. 

Our configuration looks like:
![image](https://github.com/MarcGiffing/bucket4j-spring-boot-starter/assets/10775683/4f62c784-b106-4d24-a6d3-f1cc1582f7f2)

Everything works perfectly and out-of-the-box. But we want to add alerts sent from Grafana based on the reaching 50% of the limit. We also have prometheus configured. So we would like to expose the metrics which show the current load of each API. 

Currently the only metrics we get in the prometheus is bucket4j_summary_consumed_total per each tag, but we need not the total but the amount of consumed buckets per second and per minute to understand how much buckets were consumed and how much are left. This would allow us (with knowing the bandwith capacity) to create alerts. 
But as far as I understand, there are no other default metrics, so I assume I have to write custom filter and somehow calculate the current load of APIs. 

I have found the FilterConfiguration class which holds the collection of rate limit checks which I could re-use in my custom filter to analyze the amount of remaining and consumed buckets but I can't access it:
`private List<RateLimitCheck<R>> rateLimitChecks = new ArrayList<>();`

Questions which I have:

1) Is it possible to mention custom filter in the bucket4j yml config so it will be called only for the APIs mentioned in the configuration instead of all APIs? 
2) Is it possible to fetch the current consumed/rejected buckets amount in this filter?

Any help appreciated, thank you!

**Comment by MarcGiffing** [comment]
I'm not sure but can you use something like this:
`increase(bucket4j_summary_consumed_total[1m])`
And then add an alert if the value is above 250. If you add the EXTERNAL_SERVICES_MAX_REQUESTS_PER_MINUTE as a tag you can maybe make it more dynamically.
https://torstenmandry.github.io/Prometheus-Counters/

But all customers in your configuration use the same rate limit key. If you want to rate limit independently you should consider to add a cache-key.

If you want to implement you own metric handler you can provide a bean which implements MetricHandler. It will automatically be picked up.

```java
@Component
public class DebugMetricHandler implements MetricHandler {

	@Override
	public void handle(MetricType type, String name, long tokens, List<MetricTagResult> tags) {
          // your code here
	}

}
```

**Reply by MarcGiffing** [reply]
You should also take into account that the refill speed is per default greedy, that means that the tokens are refilled as soon as possible.

**Reply by dmitry-naumovich** [reply]
Thank you, yes, I understand the it's greedy but anyway I hope to see the current state of buckets. I will try to use the solutions you've suggested and then get back here.
